subprojects/core/src/main/java/org/gradle/internal/operations/logging/OutputDetailsFactory.java

```diff
@@ -0,0 +1,134 @@
+/*
+ * Copyright 2018 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.internal.operations.logging;
+
+import org.gradle.internal.logging.buildoperation.LogEventBuildOperationProgressDetails;
+import org.gradle.internal.logging.buildoperation.ProgressStartBuildOperationProgressDetails;
+import org.gradle.internal.logging.buildoperation.StyledTextBuildOperationProgressDetails;
+import org.gradle.internal.logging.events.CategorisedOutputEvent;
+import org.gradle.internal.logging.events.LogEvent;
+import org.gradle.internal.logging.events.ProgressStartEvent;
+import org.gradle.internal.logging.events.StyledTextOutputEvent;
+
+import java.util.List;
+
+public final class OutputDetailsFactory {
+    public static Object from(CategorisedOutputEvent event) {
```

Comment:
We should avoid this object allocation and just use the underlying types. They are internal, so we can just make them implement the interfaces directly.